File: backend/app/api/articles.py
from __future__ import annotations

from typing import Annotated, Literal

from fastapi import APIRouter, HTTPException, Query

from app.services.article_reader import article_detail, article_summary, get_article, list_articles

router = APIRouter()


@router.get("/articles")
def articles(
    q: Annotated[str | None, Query(max_length=200)] = None,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 20,
    category: Annotated[str | None, Query(max_length=200)] = None,
    sort: Annotated[Literal["date_desc", "archive_desc", "title_asc", "relevance"] | None, Query()] = None,
) -> dict[str, object]:
    query = q.strip() if q and q.strip() else None
    selected_category = category.strip() if category and category.strip() else None
    effective_sort = sort or ("relevance" if query else "date_desc")
    matched = list_articles(query, category=selected_category, sort=effective_sort)
    total = len(matched)
    total_pages = (total + page_size - 1) // page_size
    start = (page - 1) * page_size
    items = [article_summary(article) for article in matched[start : start + page_size]]
    return {
        "items": items,
        "total": total,
        "query": query,
        "category": selected_category,
        "sort": effective_sort,
        "page": page,
        "page_size": page_size,
        "total_pages": total_pages,
        "has_next": page < total_pages,
        "has_previous": total_pages > 0 and page > 1,
    }


@router.get("/articles/{article_id}")
def article(article_id: str) -> dict[str, object]:
    stored_article = get_article(article_id)
    if stored_article is None:
        raise HTTPException(status_code=404, detail="Article not found")
    return article_detail(stored_article)

File: frontend/src/lib/articles.ts
export type ArticleMetadata = {
  date?: string | null;
  category?: string | null;
  references?: Array<string | Record<string, unknown>>;
  images?: Array<string | Record<string, unknown>>;
};

export type ArticleListSort = "date_desc" | "archive_desc" | "title_asc" | "relevance";

export type ArticleListRequest = {
  q?: string;
  page?: number;
  page_size?: number;
  category?: string;
  sort?: ArticleListSort;
};

export type ArticleSummary = {
  id: string;
  title: string;
  url: string;
  metadata: ArticleMetadata;
  content_preview: string;
};

export type ArticleDetail = {
  id: string;
  title: string;
  url: string;
  content: string;
  metadata: ArticleMetadata;
};

export type ArticleListResponse = {
  items: ArticleSummary[];
  total: number;
  query: string | null;
  category: string | null;
  sort: ArticleListSort;
  page: number;
  page_size: number;
  total_pages: number;
  has_next: boolean;
  has_previous: boolean;
};

const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";

export async function fetchArticles(queryOrOptions?: string | ArticleListRequest): Promise<ArticleListResponse> {
  const url = new URL("/articles", API_BASE_URL);

  const options: ArticleListRequest = typeof queryOrOptions === "string" ? { q: queryOrOptions } : queryOrOptions ?? {};
  if (options.q?.trim()) {
    url.searchParams.set("q", options.q.trim());
  }
  if (Number.isFinite(options.page) && (options.page as number) > 0) {
    url.searchParams.set("page", String(options.page));
  }
  if (Number.isFinite(options.page_size) && (options.page_size as number) > 0) {
    url.searchParams.set("page_size", String(options.page_size));
  }
  if (options.category?.trim()) {
    url.searchParams.set("category", options.category.trim());
  }
  if (options.sort) {
    url.searchParams.set("sort", options.sort);
  }

  const response = await fetch(url.toString(), { cache: "no-store" });
  if (!response.ok) {
    throw new Error(`Failed to load articles: ${response.status}`);
  }
  return response.json() as Promise<ArticleListResponse>;
}

export async function fetchArticle(articleId: string): Promise<ArticleDetail> {
  const response = await fetch(new URL(`/articles/${articleId}`, API_BASE_URL).toString(), {
    cache: "no-store",
  });
  if (response.status === 404) {
    throw new Error("Article not found");
  }
  if (!response.ok) {
    throw new Error(`Failed to load article: ${response.status}`);
  }
  return response.json() as Promise<ArticleDetail>;
}

export function formatMetadata(metadata: ArticleMetadata): string {
  const parts = [metadata.date, metadata.category].filter(Boolean);
  return parts.length ? parts.join(" · ") : "No metadata";
}
